### PR TITLE
Allow consumer ops to sync on GraphRoot's gradient

### PR DESCRIPTION
**Pull request by mcarilli**
Currently, a GraphRoot instance doesn't have an associated stream.  Streaming backward synchronization logic assumes the instance ran on the default stream, and tells consumer ops to sync with the default stream.  If the gradient the GraphRoot instance passes to consumer backward ops was populated on a non-default stream, we have a race condition.

The race condition can exist even if the user doesn't give a manually populated gradient:
```python
with torch.cuda.stream(side_stream):
    # loss.backward() implicitly synthesizes a one-element 1.0 tensor on side_stream
    # GraphRoot passes it to consumers, but consumers first sync on default stream, not side_stream.    
    loss.backward()

    # Internally to backward(), streaming-backward logic takes over, stuff executes on the same stream it ran on in forward,
    # and the side_stream context is irrelevant.  GraphRoot's interaction with its first consumer(s) is the spot where
    # the side_stream context causes a problem.
```

This PR fixes the race condition by associating a GraphRoot instance, at construction time, with the current stream(s) on the device(s) of the grads it will pass to consumers. (i think this relies on GraphRoot executing in the main thread, before backward thread(s) fork, because the grads were populated on the main thread.)

The test demonstrates the race condition. It fails reliably without the PR's GraphRoot diffs and passes with the GraphRoot diffs.

With the GraphRoot diffs, manually populating an incoming-gradient arg for `backward` (or `torch.autograd.grad`) and the actual call to `autograd.backward` will have the same stream-semantics relationship as any other pair of ops:
```python
# implicit population is safe
with torch.cuda.stream(side_stream):
    loss.backward()

# explicit population in side stream then backward in side stream is safe
with torch.cuda.stream(side_stream):
    kickoff_grad = torch.ones_like(loss)
    loss.backward(gradient=kickoff_grad)

# explicit population in one stream then backward kickoff in another stream
# is NOT safe, even with this PR's diffs, but that unsafety is consistent with
# stream-semantics relationship of any pair of ops
kickoff_grad = torch.ones_like(loss)
with torch.cuda.stream(side_stream):
    loss.backward(gradient=kickoff_grad)

# Safe, as you'd expect for any pair of ops
kickoff_grad = torch.ones_like(loss)
side_stream.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(side_stream):
    loss.backward(gradient=kickoff_grad)
```
This PR also adds the last three examples above to cuda docs and references them from autograd docstrings.

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -1733,6 +1733,46 @@ def test_streaming_backwards_device_transfer(self):
...
         self.assertTrue(b.grad.sum().item() == 4 * size)
 
+    def test_streaming_backward_sync_graph_root(self):
+        # This function tests if bwd ops running on a side stream properly sync with the graph root.
+        fwd_bwd_op_stream = torch.cuda.Stream()
+        bwd_ambient_stream = torch.cuda.Stream()
+        # We need these streams to be different otherwise the test is meaningless.
+        assert fwd_bwd_op_stream != bwd_ambient_stream
+
+        size = int(1e5)
```
Comment (albanD) [inline test/test_cuda.py:1743]: Why do you need such a big size here? To make sure that the ops are slow enough that you don't have flakyness?

File: test/test_cuda.py [hunk 2]
```diff
@@ -1733,6 +1733,46 @@ def test_streaming_backwards_device_transfer(self):
...
+
+        a = torch.full((size,), 2.0, device="cuda", requires_grad=True)
+        b = torch.full((size,), 3.0, device="cuda", requires_grad=True)
+
+        # I don't think we need any manual record_streams below.
+        # a and b remain in scope for the entire test.
+        # c and grad remain in scope for each iteration, and there's a full sync between iterations.
+        for trial in range(5):
```
Comment (albanD) [inline test/test_cuda.py:1751]: Why 5 here?
I guess there is some flakyness to this test where it might miss failures?

File: test/test_cuda.py [hunk 3]
```diff
@@ -1733,6 +1733,46 @@ def test_streaming_backwards_device_transfer(self):
...
+                grad = torch.full((size,), float(trial), device="cuda")
+
+                # Bwd ops still run on fwd_bwd_ops_stream, so the following ensures
+                # bwd ops sync properly with bwd_ambient_stream before consuming grad.
+                torch.autograd.backward(tensors=c, grad_tensors=grad)
+
+                # assertEquals below run on bwd_ambient_stream, so this test may also fail
+                # if backward() fails to sync with the ambient stream at the end.
+                # A torch.cuda.synchronize() here would ensure this test focuses on GraphRoot only.
+                # Which do we prefer:  A focused test, or a test that has a chance to catch additional related bugs?
```
Comment (albanD) [inline test/test_cuda.py:1771]: Testing both sounds better to me especially if there is almost no additional cost.

File: test/test_cuda.py [hunk 4]
```diff
@@ -1733,6 +1733,46 @@ def test_streaming_backwards_device_transfer(self):
         self.assertTrue(a.grad.sum().item() == 4 * size)
         self.assertTrue(b.grad.sum().item() == 4 * size)
 
+    def test_streaming_backward_sync_graph_root(self):
```
Comment (albanD) [inline test/test_cuda.py:1736]: How long is this test? Should be we consider a `@slowTest` here?

**Review by albanD — COMMENTED** [review]
Very good catch! Thanks for the fix and the doc update!

File: test/test_cuda.py (hunk 3 above)
Comment (mcarilli) [reply test/test_cuda.py:1771]: Will leave as is.

File: test/test_cuda.py (hunk 4 above)
Comment (mcarilli) [reply test/test_cuda.py:1736]: It's quick, 2 seconds when i run in isolation and profiling shows the vast majority is cuda context initialization.  No need for @slowTest imo.

File: test/test_cuda.py (hunk 2 above)
Comment (mcarilli) [reply test/test_cuda.py:1751]: The bug is a race condition, likely but not guaranteed to surface in any one trial, so i want to run several trials.

Comment (albanD) [reply test/test_cuda.py:1751]: So this test can flakily pass but all failures are legitimate right?
I think it would be nice to add a comment at the top explaining this. So that if we see it being flaky in the future, we know that it should not just be disabled and assumed to be false positive.

File: test/test_cuda.py (hunk 1 above)
Comment (mcarilli) [reply test/test_cuda.py:1743]: That was my first motivation, but it turned out adjusting the size wasn't very effective at exposing the race.  The [`torch.cuda._sleep(int(50 * get_cycles_per_ms()))`](https://github.com/pytorch/pytorch/pull/45787/files#diff-e68a7bc6ba14f212e5e7eb3727394b40R1760) turned out to expose the race much more effectively, but since the runtime of the test depended negligibly on `size`, i didn't change `size`.  I'll reduce it to 1e3, no reason to keep 1e5 now.  For the record I don't think 1e5 matters much for end to end runtime or memory use.

Comment (albanD) [reply test/test_cuda.py:1743]: Yes I'm sure there is not big reason but we have many many tests. So if all of them use small sizes, it ends up making a visible difference :)

File: test/test_cuda.py (hunk 2 above)
Comment (mcarilli) [reply test/test_cuda.py:1751]: > So this test can flakily pass but all failures are legitimate right?

Yes, but the multiple trials and `torch.cuda._sleep(int(50 * get_cycles_per_ms()))` make flaky passes unlikely.

Comment (albanD) [reply test/test_cuda.py:1751]: Sounds good!
A comment explaining that would be great :)

File: test/test_cuda.py (hunk 1 above)
Comment (mcarilli) [reply test/test_cuda.py:1743]: changed to 1e3

File: test/test_cuda.py (hunk 2 above)
Comment (mcarilli) [reply test/test_cuda.py:1755]: hopefully resolved

**Review by albanD — APPROVED** [review]
LGTM thanks for the update!

**Comment by jeffdaily** [comment]
@albanD and @mcarilli, if you could lend your insight.  The new `test_streaming_backward_sync_graph_root` is flaky on our ROCm CI.  Using a host similar to our CI hosts, I can get it to somewhat reliably fail (approximately 1 out of 5 times) if I run the test in a loop while also running the `stress` command line tool.

I'd like to know if this indicates a problem in our runtime, or a problem in pytorch.  This is the line that reliably fails:
https://github.com/pytorch/pytorch/blob/7731370e718ad16994d8da25fbdc9de64a22ec44/test/test_cuda.py#L1774

The `assertEqual` fails because it eventually runs `torch.allclose()` on `a.grad` and `grad * b`.  `a.grad` is not None (I have asserted), so it seems to me that `a.grad` hasn't yet been updated by the time `grad * b` completes and the comparison kenrels start.  It seems there might be a missing cross-stream sync somewhere.

**Comment by mcarilli** [comment]
@jeffdaily the test might be exposing https://github.com/pytorch/pytorch/pull/45787/files#diff-893b1eea27352f336f4cd832919e48d721e4e90186e63400b8596db6b82e7450R1772-R1773.

Can you put a torch.cuda.synchronize() right after backward() and see if that helps?

**Comment by jeffdaily** [comment]
@mcarilli That seems to help.  How should we proceed?  File a new issue?  Do you have an idea of how to fix this already?

**Comment by jeffdaily** [comment]
Created new issue. https://github.com/pytorch/pytorch/issues/47028
